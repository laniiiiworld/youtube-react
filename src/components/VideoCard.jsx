import React from 'react';
import { formatAgo } from '../util/date';

export default function VideoCard({
  video: {
    snippet: { title, thumbnails, channelTitle, publishedAt },
  },
}) {
  return (
    <li className='cursor-pointer flex flex-col'>
      <img src={thumbnails.medium.url} alt={title} className='rounded-xl w-full' />
      <div>
        <p className='font-semibold my-2 line-clamp-2'>{title}</p>
        <p className='text-sm opacity-80'>{channelTitle}</p>
        <span className='text-sm opacity-80'>{formatAgo(publishedAt)}</span>
      </div>
    </li>
  );
}
