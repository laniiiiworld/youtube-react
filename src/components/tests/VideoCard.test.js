import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route, useLocation } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withRouter } from '../../tests/utils';
import { fakeVideo as video } from '../../tests/videos';
import VideoCard from '../VideoCard';

describe('VideoCard', () => {
  it('renders grid type correctly', () => {
    const component = renderer.create(
      withRouter(<Route path='/' element={<VideoCard video={video} type='grid' />} />) //
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders list type correctly', () => {
    const component = renderer.create(
      withRouter(<Route path='/' element={<VideoCard video={video} type='list' />} />) //
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('navigates to detailed video page with video state when clicked', () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }
    render(
      withRouter(
        <>
          <Route path='/' element={<VideoCard video={video} />} />
          <Route path={`/videos/watch/${video.id}`} element={<LocationStateDisplay />} />
        </>
      )
    );

    const card = screen.getByRole('listitem');
    userEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
