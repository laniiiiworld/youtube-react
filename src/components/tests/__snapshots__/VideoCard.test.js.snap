// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<li
  className="cursor-pointer flex flex-col"
  onClick={[Function]}
>
  <img
    alt="title"
    className="rounded-xl w-full"
    src="http://image/"
  />
  <div
    className=""
  >
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-sm opacity-80"
    >
      channel title
    </p>
    <span
      className="text-sm opacity-80"
    >
      just now
    </span>
  </div>
</li>
`;

exports[`VideoCard renders list type correctly 1`] = `
<li
  className="cursor-pointer flex flex-row mb-2"
  onClick={[Function]}
>
  <img
    alt="title"
    className="rounded-xl w-5/12 mr-2"
    src="http://image/"
  />
  <div
    className="w-7/12"
  >
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-sm opacity-80"
    >
      channel title
    </p>
    <span
      className="text-sm opacity-80"
    >
      just now
    </span>
  </div>
</li>
`;
